boot/setefibootvars_linux.go

```diff
@@ -0,0 +1,197 @@
+		}
+		usedBootNums[varNumber] = true
+	}
+	for bootNum := uint64(0); bootNum <= 0xFFFF; bootNum++ {
+		if !usedBootNums[bootNum] {
+			return uint16(bootNum), ErrNoMatchingVariable
+		}
+	}
+	return 0, ErrAllBootNumsUsed
+}
+
+// Ensures that a Boot#### variable contains the given EFI load option.
+//
+// It may be the case that an existing boot variable already contains the
+// given load option, in which case that boot variable is reused. Otherwise,
+// finds the first unused boot variable number and uses it. Writes the load
+// option to that variable, and returns the number of the variable that was
+// used.
+func setEfiBootOptionVariable(loadOptionData []byte) (uint16, error) {
+	bootNum, err := findMatchingBootOption(loadOptionData)
+	if err == nil {
+		return bootNum, nil
+	} else if err != ErrNoMatchingVariable {
+		return 0, err
+	}
+	varName := fmt.Sprintf("Boot%04X", bootNum)
+	err = efiWriteVariable(varName, efi.GlobalVariable, defaultVarAttrs, loadOptionData)
+	return bootNum, err
+}
+
+// Reads the current BootOrder variable, inserts the given newBootNum at the
+// beginning of the number list (and removes it from later in the list if
+// it occurs) and writes the list as the new BootOrder variable.
+func setEfiBootOrderVariable(newBootNum uint16) error {
+	origData, attrs, err := efiReadVariable("BootOrder", efi.GlobalVariable)
+	if err == efi.ErrVarNotExist {
+		attrs = defaultVarAttrs
+		origData = make([]byte, 0)
+	} else if err != nil {
+		return err
+	}
+	if len(origData)%2 != 0 {
+		return ErrInvalidBootOrder
+	}
+	var optionOffset = -1
+	for i := 0; i < len(origData); i += 2 {
+		bootNum := binary.LittleEndian.Uint16(origData[i : i+2])
+		if newBootNum == bootNum {
+			optionOffset = i
+			break
+		}
+	}
+	var newData []byte
+	if optionOffset == 0 {
+		// newBootNum already at start, no need to re-write variable
+		return nil
+	} else if optionOffset == -1 {
+		// newBootNum not in original boot order
+		newData = make([]byte, len(origData)+2)
+		binary.LittleEndian.PutUint16(newData, newBootNum)
+		copy(newData[2:], origData)
+	} else {
+		newData = make([]byte, len(origData))
+		binary.LittleEndian.PutUint16(newData, newBootNum)
+		copy(newData[2:], origData[:optionOffset])
+		copy(newData[optionOffset+2:], origData[optionOffset+2:])
+	}
+	return efiWriteVariable("BootOrder", efi.GlobalVariable, attrs, newData)
+}
+
+// SetEfiBootVariables sets the Boot#### and BootOrder variables for the given
+// load option information.
+//
+// Constructs an EFI load option with the given description, the device path
+// corresponding to the given asset path, and the given optional data. Writes
+// the EFI boot variable Boot#### to contain the resulting load option. Then,
+// sets the BootOrder variable so that the #### number from the chosen Boot####
+// is first, removing it from elsewhere in the BootOrder if it occurs.
+func SetEfiBootVariables(description string, assetPath string, optionalData []byte) error {
```

Comment:
this is exported mostly for testing via spread tests? maybe the doc should remark that is usually not meant to be called directly